### PR TITLE
Added Buildkite pipeline and integration tests

### PR DESCRIPTION
**Pull request by andreeaflorescu**
This pipeline contains the tests that are currently run for
kvm-ioctls. The pipeline also has a step that runs the coverage
test defined in integration_tests/test_coverage.py.

The purpose of this pipeline is to be used by all rust-vmm
repositories to ensure we keep the same quality bar for all crates.

### CODE REVIEW

**Comment by andreeaflorescu** [comment]
The pipeline and integration tests were tested on kvm-ioctls with Buildkite: https://buildkite.com/rust-vmm/fandree-kvm-ioctls-ci and locally with linux-loader.

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -0,0 +1,176 @@
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
```
Comment (jiangliu) [inline .buildkite/pipeline.yml]: Should we also consider BSD-3-Clause?

File: README.md [hunk 2]
```diff
@@ -1 +1,172 @@
-# rust-vmm-ci
\ No newline at end of file
+# rust-vmm-ci
+
+The `rust-vmm-ci` contains integration tests and [Buildkite]() pipeline
```
Comment (aghecenco) [inline README.md]: The link is missing.

File: README.md [hunk 3]
```diff
@@ -1 +1,172 @@
...
+definitions that are used for running the CI for all rust-vmm crates.
+
+Having a centralized place for the tests is one of the enabler for keeping the
```
Comment (aghecenco) [inline README.md]:
```suggestion
Having a centralized place for the tests is an enabler for keeping the
```

File: README.md [hunk 4]
```diff
@@ -1 +1,172 @@
...
+DEB_URL="http://ftp.debian.org/debian/pool/main/l/linux/${DEB_NAME}"
+
+REPO_PATH="${BUILDKITE_BUILD_CHECKOUT_PATH}"
+DEB_PATH="${REPO_PATH}/${DEB_NAME}"
+EXTRACT_PATH="${REPO_PATH}/src/bzimage-archive"
+BZIMAGE_PATH="${EXTRACT_PATH}/boot/vmlinuz-4.9.0-9-amd64"
+
+mkdir -p ${EXTRACT_PATH}
+
+wget $DEB_URL -P ${REPO_PATH}
```
Comment (aghecenco) [inline README.md]:
```suggestion
wget ${DEB_URL} -P ${REPO_PATH}
```

File: README.md [hunk 5]
```diff
@@ -1 +1,172 @@
...
+```
+
+In this example the post-checkout hook downloads a deb image, extracts the
+contents of it and places it in `linux-loader/src/bzimage`. The unit tests
+will use the relative path `src/bzimage` which does not depend on the image
+being downloaded.
+
+## Integration Tests
+
+Besides the one-liner tests defined in the
```
Comment (aghecenco) [inline README.md]:
```suggestion
In addition to the one-liner tests defined in the
```

File: integration_tests/test_coverage.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+"""Test the coverage and update the threshold when coverage is increased."""
+
+import json, os, re, shutil, subprocess
+import pytest
+
+from utils import get_repo_root_path
+
+REPO_ROOT_PATH = get_repo_root_path()
+COVERAGE_CONFIG_PATH = os.path.join(REPO_ROOT_PATH, "coverage_config.json")
+
```
Comment (aghecenco) [inline integration_tests/test_coverage.py]: nit: 2 blank lines before top level functions

File: README.md [hunk 7]
```diff
@@ -1 +1,172 @@
...
+
+4. The code owner of the repository will have to setup a WebHook for triggering
+the CI on Pull Requests and pushes.
+
+## Buildkite Pipeline
+
+The Buildkite pipeline is the definition of integration tests to be run as part
+of the CI. It includes steps for running unit tests and linters (including
+coding style checks), and computing the coverage.
+
```
Comment (sameo) [inline README.md]: This section is a little confusing to me. It feels like there's a need to define that pipeline whereas in reality this is what the git submodule brings.
So I would refer back to step 1 of the previous section, explaining that this steps adds a buildkite pipeline into the repo and that buildkite will use it.

File: README.md [hunk 8]
```diff
@@ -1 +1,172 @@
...
+      - docker#v3.0.1:
+          image: "rustvmm/dev:v2"
+          always-pull: true
+```
+
+To see all steps in the pipeline check the
+[.buikite/pipeline.yml](.buildkite/pipeline.yml) file.
+
+### Custom Repository Hooks
+
```
Comment (sameo) [inline README.md]: We should also define a section for adding a custom buildkite pipeline. The linux-loader (again) for example needs a custom one in order to build all features combinations.

File: README.md [hunk 9]
```diff
@@ -1 +1,172 @@
...
+To run the integration tests defined in the pipeline as part of the CI:
+
+1. Add rust-vmm-ci as a git submodule to your repository
+
+```bash
+git submodule add https://github/rust-vmm/rust-vmm-ci
+git add rust-vmm-ci/
+git commit -s -m "Added rust-vmm-ci as submodule"
+```
+
```
Comment (sameo) [inline README.md]: You may want to briefly what that does and why we do this.

File: .buildkite/pipeline.yml (hunk 1 above)
Comment (andreeaflorescu) [reply .buildkite/pipeline.yml]: I am not sure if the BSD-3-Clause is needed in this case because this is not an "exportable" repository. If we do add the BSD-3-clause license as well all crates in rust-vmm need to have a copyright and attribution for this repository which is just running the tests. I find that to be an overkill, but if people think we should add it, I'll follow.

File: README.md (hunk 7 above)
Comment (andreeaflorescu) [reply README.md]: This section is not about using rust-vmm-ci, but about how we created rust-vmm-ci and what does it have. It's a "developer" readme. Should I put it in a different .md file and keep the readme only with information about how to use rust-vmm-ci?

File: README.md (hunk 9 above)
Comment (andreeaflorescu) [reply README.md]: Do you mean the submodule?

File: README.md (hunk 8 above)
Comment (andreeaflorescu) [reply README.md]: Added section: https://github.com/rust-vmm/rust-vmm-ci/blob/5d90acfbcb4166cb1e5dbd2487108a3e9aea0f79/README.md#custom-pipeline

File: README.md (hunk 7 above)
Comment (andreeaflorescu) [reply README.md]: I did some rephrasing here and there. After some more thinking I decided it's better to have only one readme for now. Maybe we should have both sections (Buildkite Pipeline and Integration Tests) in another section, but I can't find a good name. Maybe "rust-vmm-ci development"? What do you say?

File: README.md [hunk 10]
```diff
@@ -1 +1,204 @@
-# rust-vmm-ci
\ No newline at end of file
+# rust-vmm-ci
+
+The `rust-vmm-ci` contains [integration tests](#integration-tests) and
```
Comment (aghecenco) [inline README.md]: nit
```suggestion
The `rust-vmm-ci` repository contains [integration tests](#integration-tests) and
```

File: integration_tests/test_coverage.py (hunk 6 above)
Comment (aghecenco) [reply integration_tests/test_coverage.py]: I'm keeping my nit :smile:

File: README.md (hunk 9 above)
Comment (sameo) [reply README.md]: I mean to explain what does it actually do and why we're doing it. A readable explanation after the set of commands.
But that's a nit.

File: README.md [hunk 11]
```diff
@@ -1 +1,204 @@
...
+4. The code owner of the repository will have to setup a WebHook for triggering
+the CI on Pull Requests and pushes.
+
+## Buildkite Pipeline
+
+The [Buildkite](https://buildkite.com) pipeline is the definition of tests to
+be run as part of the CI. It includes steps for running unit tests and linters
+(including coding style checks), and computing the coverage.
+
+Currently the tests run on a Linux machine on x86_64 and aarch64.
```
Comment (sameo) [inline README.md]: Nit: `a` here feels like it's running on a specific machine. I'd rephrase it as:
`Currently the tests can run on Linux x86_64 and aarch64 hosts`

File: README.md (hunk 7 above)
Comment (sameo) [reply README.md]: +1 for one single readme.
This looks fine to me.

File: README.md (hunk 8 above)
Comment (sameo) [reply README.md]: Thanks.

**Comment by andreeaflorescu** [comment]
@aghecenco @jiangliu @sameo I fixed the remaining nits and a tiny bug in the bash script that was adding rust-vmm-ci as a submodule. Please take another look.
